### ISSUE TITLE
BUG? Bash export of an else-if-construct

### ISSUE THREAD

**Issue by reks99** [fesch/Structorizer.Desktop#386]
Hi,

I found, that structuricer some times generates an empty else path, when exported to bash. Only the red marked path is wrong and looks like 
```
 if [ -z "$PAGER" ]
 then
  PAGER=more
 else
  # Sollte frei bleiben.
 fi
``` 
Code has to be natively exported .

![mn-2](https://cloud.githubusercontent.com/assets/18035140/25085314/cc5361f2-2361-11e7-9660-af7d9eb66676.png)

The attached file mn-2.txt is the nsd - renamed, that github accepts the file.
When exported, the problem is around line  52.

[mn-2.txt](https://github.com/fesch/Structorizer.Desktop/files/925246/mn-2.txt)

Fine regards
Rolf

**Comment by GitMensch** [comment]
@codemanyak Can issues be deleted?

**Comment by codemanyak** [comment]
@GitMensch Apperently no, but they can be emptied and closed :wink: - this was done by the originator himself.

**Comment by reks99** [comment]
Hi,

I found, that structuricer generates an empty else path, when exported to bash.


![mn-2](https://cloud.githubusercontent.com/assets/18035140/25084597/c60117b2-235d-11e7-8e0e-ff36925546ff.png)


[mn-2.txt](https://github.com/fesch/Structorizer.Desktop/files/925246/mn-2.txt)

**Comment by codemanyak** [comment]
According to the source code, the else branch of alternatives will only be exported if it is not empty, i.e. if it contains at least one element, no matter what the element contains. So if you happen to hold an empty instruction element in the else branch then the branch is not regarded as empty.
Are you aware that the uploaded file (mn-2.nsd, disguised as ...txt) differs from the uploaded image in that not the small else branch (containing the comment "Sollte frei bleiben" but the large else branch of the enclosing alternative with exactly the same condition (`[ -z "$PAGER" ]`) is coloured red. So about which else branch are we actually talking? As soon as you place a comment in the branch, it will no longer be empty in a formal way, because it takes an (instruction) element to hold a comment. Consequently, I get an else branch in code around the comment "# Sollte frei bleiben", but for the really empty else branch of the outer alternative I don't.
May that be nature of your worrying?

**Comment by codemanyak** [comment]
Ok, obviously bash doesn't cope with empty else branches. I hadn't been aware of it. Since there is no "NOOP" instruction in the shell (like single ';' in C), I could offer to outcomment the "else" keyword in case the branch contains only algorithmically empty elements with mere comments. This way, the comments would still be associated to the (non-existent) else branch.

**Comment by codemanyak** [comment]
The same problem occurs with other algorithmic structures like loops as well. Their export causes syntax errors in bash if their body is empty.

**Comment by reks99** [comment]
Hi Kay,

the comment in the empty else path was only set, to find the element in the exported code an empty element was also generated, without a comment.

I like your idea, to comment out the empty path, if there is no algorithmic code in the path.

Fine regards
Rolf

**Comment by codemanyak** [comment]
> the comment in the empty else path was only set, to find the element in the exported code an empty element was also generated, without a comment.

I tried with an emptied branch: no "else" line is generated. I guess you will have had an empty instruction there (just the text cleared, such that no empty set symbol appeared).

> I like your idea, to comment out the empty path, if there is no algorithmic code in the path.

I liked it, too - until I realized that the same problem could occur in loops and any other structured elements (e.g. as a result of commenting out all substantial elements). In these cases the entire structure would have to be commented out. And this could recursively cause an entire tree upwards having to be commented out (though parts of it will already have been expressed).
Hence it turned out to be way easier and safer to place a colon where a no-op Subqueue occurs) - if you can also agree with it.

**Comment by codemanyak** [comment]
Will be fixed in release 3.27 (current fix uploaded to [codemanyak/Structorizer.Desktop/master](https://github.com/codemanyak/Structorizer.Desktop/archive/master.zip)).
@reks99 
Do you need a quick fix or can it wait till release 3.27 (planned for about end of May)?
(You might of course [download](https://github.com/codemanyak/Structorizer.Desktop/archive/master.zip) and compile the current source code preview from codemanyak branch.)

**Comment by reks99** [comment]
Hi Kay,

I tested the skript with a colon in the empty part and it works. 

I hope I don't need the fix imediately. 
Thanks
Rolf

**Comment by GitMensch** [comment]
As other items that were coded for https://github.com/fesch/Structorizer.Desktop/milestone/2 are closed already this can be closed, too.

> Do you need a quick fix or can it wait till release 3.27 (planned for about end of May)?

@codemanyak
Can we hope for 3.27 before/in August?
